NMS-14455: Install JDBC Driver on Core (#5113)
OpenNMS/opennms: docs/modules/deployment/pages/core/install-jdbc-driver.adoc

[[install-jdbc-driver-core]]
= Install JDBC Driver on OpenNMS {page-component-title} Core

To detect, poll, or collect any JDBC service, you must install the appropriate JDBC driver.

. Download the JDBC driver JAR file from a repository (for example, a public https://mvnrepository.com/artifact/mysql/mysql-connector-java/8.0.22[Maven repository]) or from the database vendor.

> . Install the JAR file in \$OPENNMS_HOME/lib/

Author: ```suggestion
. Install the JAR file in `\$OPENNMS_HOME/lib/`.
```

. Restart {page-component-title}

> Testing an installed JDBC Driver

Reviewer: This line doesn't make sense. I think it should be:
```suggestion
== Test an installed JDBC driver
```

.Connect to the Karaf shell
[source,console]
----
ssh -p 8101 admin@localhost
----

> .Attempt an ad-hoc collection using the database driver

Reviewer: ```suggestion
.Attempt an ad hoc collection using the database driver
```
Inspired by our recent style guide foray in to Latin, I googled "ad hoc" and discovered that it doesn't need to be hyphenated, even when it's a modifier.

[source, karaf]
----
collect org.opennms.netmgt.collectd.JdbcCollector 127.0.0.1 driver=org.mysql.jdbc.Driver collection=MySQL-Global-Stats-51 url=jdbc:mysql://OPENNMS_JDBC_HOSTNAME:3306/test
----

> .If the JDBC driver is not available or the driver classname is incorrect, the collect command will return an error:

Reviewer: ```suggestion
.If the JDBC driver is not available or the driver classname is incorrect, the collect command returns an error:
```

[[install-jdbc-driver-core]]

> = Install JDBC Driver on OpenNMS {page-component-title} Core

Reviewer 2: ```suggestion
= Install JDBC Driver on {page-component-title} Core
```

To detect, poll, or collect any JDBC service, you must install the appropriate JDBC driver.

> . Download the JDBC driver JAR file from a repository (for example, a public https://mvnrepository.com/artifact/mysql/mysql-connector-java/8.0.22[Maven repository]) or from the database vendor.

Reviewer 2: ```suggestion
. Download the JDBC driver JAR file from a repository (for example, a public https://mvnrepository.com/artifact/mysql/mysql-connector-java/8.0.30[Maven repository]) or from the database vendor.
```

> . Install the JAR file in `\$OPENNMS_HOME/lib/`.

Reviewer 2: ```suggestion
. Install the JAR file in `$\{OPENNMS_HOME}/lib/`.
```

[[install-jdbc-driver-core]]
= Install JDBC Driver on OpenNMS {page-component-title} Core

> To detect, poll, or collect any JDBC service, you must install the appropriate JDBC driver.
> 
> We reference the repository directory relative to the {page-component-title} home directory.
> The home directory will differ based on your operating system.
> // These directories need to be updated; I pulled them from the Install JDBC Driver on Minion page.
> For Debian and Ubuntu, the home directory is `usr/share/minion`.
> For CentOS and RHEL, the home directory is `/opt/minion`.
> 
> . Download the driver JAR file from a repository (for example, a public https://mvnrepository.com/artifact/mysql/mysql-connector-java/8.0.22[Maven repository]).
> . Install the JAR file in the repository directory, following the Maven repository pattern.
> +
> For MySQL 8.0.22, the path would be `repositories/default/mysql/mysql-connector-java/8.0.22/mysql-connector-java-8.0.22.jar`.
> 
> . Create a Karaf feature for the MySQL JDBC driver.
> 
> .Create a features configuration file
> [source, console]
> ----
> sudo vi deploy/jdbc-mysql.xml
> ----
> 
> .Add the following content to download the MySQL JDBC driver
> [source, jdbc-mysql.xml]
> [subs="verbatim,attributes"]
> -----
> <?xml version="1.0" encoding="UTF-8"?>
> <features name="opennms-{page-component-version}"<1>
>           xmlns="http://karaf.apache.org/xmlns/features/v1.4.0"
>           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
>           xsi:schemaLocation="http://karaf.apache.org/xmlns/features/v1.4.0 http://karaf.apache.org/xmlns/features/v1.4.0">
>   <feature name="mysql-bundle" version="8.0.22" install="auto"><2>
>     <bundle>wrap:mvn:mysql/mysql-connector-java/8.0.22</bundle><3>
>   </feature>
> </features>
> -----
> 
> . The version should match your {page-component-title} Core version.
> . The bundle is named `mysql-bundle` and the version number is set to `8.0.22`.
> . Bundle to download version `8.0.22` with Maven from official repositories.
> 
> As soon you write the file, Karaf picks up the feature and tries to install it.
> Troubleshoot the feature installation in the `data/log/karaf.log`.
> 
> .Connect to the Karaf shell
> [source,console]
> ----
> ssh -p 8201 admin@localhost
> ----
> 
> .Verify feature installation and startup
> [source, karaf]
> ----
> feature:list | grep mysql
> ----
> 
> .MySQL bundle is installed and available
> [source, output]
> ----
> mysql-bundle   | 8.0.22   | x        | Started     | opennms-{page-component-version}  |
> ----

Reviewer 3: This is actually a lot easier on core.

```suggestion
[[install-jdbc-driver-core]]
= Install JDBC Driver on OpenNMS {page-component-title} Core

To detect, poll, or collect any JDBC service, you must install the appropriate JDBC driver.

. Download the JDBC driver JAR file from a repository (for example, a public https://mvnrepository.com/artifact/mysql/mysql-connector-java/8.0.22[Maven repository]) or from the database vendor.
. Install the JAR file in \$OPENNMS_HOME/lib/
. Restart {page-component-title}

Testing an installed JDBC Driver
.Connect to the Karaf shell
[source,console]
----
ssh -p 8101 admin@localhost
----

.Attempt an ad-hoc collection using the database driver
[source, karaf]
----
collect org.opennms.netmgt.collectd.JdbcCollector 127.0.0.1 driver=org.mysql.jdbc.Driver collection=MySQL-Global-Stats-51 url=jdbc:mysql://OPENNMS_JDBC_HOSTNAME:3306/test
----

.If the JDBC driver is not available or the driver classname is incorrect, the collect command will return an error:
[source, karaf]
----
Collect failed with: java.util.concurrent.ExecutionException: java.lang.RuntimeException: Unable to load driver class: java.lang.ClassNotFoundException: org.mariadb.jdbc.Driver 
----
```

Author: I didn't know you could offer a full-document comment, but I love it. 😂 Thanks so much!
